Feature/composition : 프로젝트 기본 구성

pray92 opened the pull request:
1. local, dev, production 프로파일 구성
2. 공통 예외 핸들러 구성
3. Functional Architecture 패키징
4. Swagger 추가
5. MyBatis / JPA 리포지토리 개별 구현(초안 구현 상태, 추후 구현하면서 개선 예정)

pray92 committed:
FEAT : 에러 코드 생성

pray92 committed:
FEAT : 공통 예외 처리 핸들러 클래스 구현

pray92 committed:
FEAT : 비즈니스 로직 및 기타 예외 핸들링 추가

pray92 committed:
FEAT : 공통 예외 핸들링 테스트를 위한 User 도메인 기능 구현

pray92 committed:
FEAT : 공통 예외 핸들링 테스트 케이스 구동

pray92 committed:
CHORE : Member -> User 네이밍 수정

pray92 committed:
FEAT : 테스트 환경 프로파일 작성 및 로컬 환경 프로파일 설정 추가

pray92 committed:
Merge remote-tracking branch 'origin/develop' into feature/composition

pray92 committed:
CHORE : 오류 메시지 수정

pray92 committed:
CHORE : DAO 반환 타입 Optional 변경

pray92 committed:
FEAT : Swagger 추가

pray92 committed:
FEAT : MyBatis 기반 유저 조회 기능 구현

f-lab-nolan commented on src/main/java/com/texturized/muus/common/error/ErrorCode.java:
```diff
@@ -9,16 +9,16 @@
-    INVALID_INPUT_VALUE(400, "C001", " invalid Input Value"),
-    METHOD_NOT_ALLOWED(405, "C002", " Entity Not Found"),
+    INVALID_INPUT_VALUE(400, "C001", " Invalid Input Value"),
+    METHOD_NOT_ALLOWED(405, "C002", " Invalid Input Value"),
```
메시지가 12라인과 같네요... 
METHOD_NOT_ALLOWED 라면 해당 메시지와는 어울리지 않아보여요.

pray92 replied:
Unsupport HTTP Method 메시지로 변경했습니다

f-lab-nolan commented on src/main/java/com/texturized/muus/common/error/exception/BusinessException.java:
```diff
@@ -0,0 +1,21 @@
+
+    private ErrorCode errorCode;
+
+    public BusinessException(String message, ErrorCode errorCode) {
```
매개변수에 final을 붙이는 습관을 들이는 게 좋아요

pray92 replied:
기본을 놓치고 있었습니다...반영했습니다

f-lab-nolan commented on src/main/java/com/texturized/muus/domain/user/dao/UserFindDao.java:
```diff
@@ -0,0 +1,25 @@
+    public Optional<User> findById(final Long id) {
+        final Optional<User> user = userRepository.findById(id);
+        user.orElseThrow(() -> new UserNotFoundException(id));
+        return user;
```
return userRepository.findById(id).orElseThrow(() -> new UserNotFoundException(id));
줄여서 가능할것 같아보이네요.

pray92 replied:
변경했습니다

f-lab-nolan commented on src/main/resources-env/local/application-local.yml:
```diff
@@ -6,10 +6,20 @@ spring:
+        dialect: org.hibernate.dialect.H2Dialect
+        format_sql: true
+        use_sql_comments: true
+        ddl-auto: create-drop
```
서버를 기동할때 마다 테이블 생성이 될텐데요.
로컬에서  JUint 테스트 말고 서버 기동해서 로컬에서 API 테스트 할때마다,
사용자 가입하고 하는 절차가 계속 있어야 할것 같은데요?

pray92 replied:
해당 부분은 좀더 고민해보겠습니다...우선 로컬에서 개발 시 테이블이 생성 및 레코드 추가할 수 있는 sql 파일도 추가제작해보겠습니다

pray92 replied:
로컬에선 `create-drop` 속성을 적용하는 방식을 선택했습니다. 그 이유는 개발 과정에서 테이블 변동은 항상 있을 것이라 판단이 되었습니다. 그렇기에 수정된 스키마 기반으로 즉시 개발 및 테스트를 통해 생산성을 늘리기 위해 선택했습니다.

스키마(유지 또는 변동된) 기반으로 데이터를 추가하는 것은 `resources` 경로에 `import.sql`에서 작업하도록 구상했습니다. 이를 통해 로컬에서 API 테스트할 때마다 필요한 데이터를 추가하는 작업을 자동화합니다.

해당 이슈를 고민하고 결정하는 과정에서, '로컬은 이렇게 하는데, 프로덕션 환경은 어떻게 할것인가?'라는 고민이 있었습니다. 프로덕션 환경 DB의 테이블 구성은 `create`, `create-drop` 같은 속성으로 두면 기존 데이터가 유실되는데다, 테이블 내 튜플 변경을 보수적으로 접근해야 합니다. 그래서 이부분은 자동화하지 않고, 로컬 및 개발 환경에서 테스트 수행 후 직접 DB에 접속해 스키마를 수정하는 방향으로 진행하겠습니다.

pray92 committed:
FIX : Method Not Allowed 에러 메시지 수정

pray92 committed:
CHORE : 필드 final 키워드 추가

pray92 committed:
CHORE : 코드 리팩토링

pray92 committed:
FEAT : 스키마 관련 파일 추가

- create-drop 이유: 개발 환경에서 스키마는 얼마든지 변동 가능성이 있다 판단, 수정된 스키마 기반으로 즉시 적용해 개발 및 테스트를 즉시 처리하기 위해 선택
- 기본적인 데이터는 resources 경로 내 import.sql에 쿼리문을 추가하여 구성, 현재는 빈 파일이나 이후 기능 구현할 때 테스트를 위해 추가하는 방식으로 진행

pray92 committed:
REFACTOR : DDD Layered Architecture 구성

pray92 commented:
DDD Layered Architecture 기반으로 패키징을 재구성해봤습니다.

크게 아래와 같이 구성을 해보았습니다.
- common
- presentation
- application
- domain

pray92 committed:
REFACTOR : 테스트 패키징 재구성
